src/compute/src/sink/persist_sink.rs

```diff
@@ -562,6 +567,112 @@ impl<T, R> CorrectionBuffer<T, R> {
     }
 }
 
+/// A stash for storing future updates by time.
+#[derive(Default)]
+struct UpdateStash<D>(BTreeMap<Timestamp, ConsolidatingVec<D>>);
+
+impl<D: Ord> UpdateStash<D> {
+    fn new() -> Self {
+        Self(Default::default())
+    }
+
+    /// Insert a batch of updates into the stash.
+    fn insert(&mut self, mut updates: Vec<(D, Timestamp, Diff)>) {
+        consolidate_updates(&mut updates);
+        updates.sort_unstable_by_key(|(_, time, _)| *time);
+
+        let mut updates = updates.into_iter().peekable();
+        while let Some(&(_, time, _)) = updates.peek() {
+            let data = updates
+                .peeking_take_while(|(_, t, _)| *t == time)
+                .map(|(d, _, r)| (d, r));
+
+            use std::collections::btree_map::Entry;
+            match self.0.entry(time) {
+                Entry::Vacant(entry) => {
+                    entry.insert(data.collect());
+                }
+                Entry::Occupied(mut entry) => {
+                    entry.get_mut().extend(data);
+                }
+            }
+        }
+    }
+
+    /// Remove all updates before the given frontier and return an iterator over them.
+    fn drain_before(
+        &mut self,
+        frontier: &Antichain<Timestamp>,
+    ) -> impl Iterator<Item = (D, Timestamp, Diff)> {
+        let retain = match frontier.as_option() {
+            Some(time) => self.0.split_off(time),
+            None => BTreeMap::new(),
+        };
+        let drain = std::mem::replace(&mut self.0, retain);
+        drain
+            .into_iter()
+            .flat_map(|(t, data)| data.into_iter().map(move |(d, r)| (d, t, r)))
+    }
+}
+
+/// A vector that consolidates its contents.
+///
+/// The vector is filled with updates until it reaches capacity. At this point, the updates are
+/// consolidated to free up space. This process repeats until the consolidation recovered less than
+/// half of the vector's capacity, at which point the capacity is doubled.
+struct ConsolidatingVec<D> {
+    inner: Vec<(D, Diff)>,
+}
+
+impl<D: Ord> ConsolidatingVec<D> {
+    fn push(&mut self, item: (D, Diff)) {
```

Comment:
Comment please :) "Pushes `item` into the vector. If the vector does not have sufficient capacity, we try to consolidate and/or double the vector's capacity. The cost of this function is $O(n \log n)$ in the number of items the vector stores."